crates/edr_solidity/src/compiler.rs

```diff
@@ -0,0 +1,753 @@
+                    let Some(contract_type) = node["contractKind"]
+                        .as_str()
+                        .and_then(|k| ContractKind::from_str(k).ok())
+                    else {
+                        continue;
+                    };
+
+                    let contract_abi =
+                        compiler_output
+                            .contracts
+                            .get(source_name)
+                            .and_then(|contracts| {
+                                contracts
+                                    .get(node["name"].as_str().unwrap())
+                                    .map(|contract| &contract.abi)
+                            });
+
+                    let (contract_id, contract) = process_contract_ast_node(
+                        file,
+                        node,
+                        contract_type,
+                        &sources,
+                        &mut contract_id_to_linearized_base_contract_ids,
+                        contract_abi.map(Vec::as_slice),
+                    )?;
+
+                    contract_id_to_contract.insert(contract_id, contract);
+                }
+                // top-level functions
+                "FunctionDefinition" => {
+                    process_function_definition_ast_node(node, &sources, None, file, None)?;
+                }
+                _ => {}
+            }
+        }
+    }
+
+    apply_contracts_inheritance(
+        &contract_id_to_contract,
+        &contract_id_to_linearized_base_contract_ids,
+    );
+
+    Ok(BuildModel {
+        file_id_to_source_file: sources,
+        contract_id_to_contract,
+    })
+}
+
+fn apply_contracts_inheritance(
+    contract_id_to_contract: &IndexMap<u32, Rc<RefCell<Contract>>>,
+    contract_id_to_linearized_base_contract_ids: &HashMap<u32, Vec<u32>>,
+) {
+    for (cid, contract) in contract_id_to_contract {
+        let mut contract = contract.borrow_mut();
+
+        let inheritance_ids = &contract_id_to_linearized_base_contract_ids[cid];
+
+        for base_id in inheritance_ids {
+            let base_contract = contract_id_to_contract.get(base_id);
+
+            let base_contract = match base_contract {
+                Some(base_contract) => base_contract,
+                // This list includes interface, which we don't model
+                None => continue,
+            };
+
+            if cid != base_id {
+                let base_contract = &base_contract.borrow();
+                contract.add_next_linearized_base_contract(base_contract);
+            }
+        }
+    }
+}
+
+fn process_contract_ast_node(
+    file: &RefCell<SourceFile>,
+    contract_node: &serde_json::Value,
+    contract_type: ContractKind,
+    sources: &Rc<BuildModelSources>,
+    contract_id_to_linearized_base_contract_ids: &mut HashMap<u32, Vec<u32>>,
+    contract_abi: Option<&[ContractAbiEntry]>,
+) -> anyhow::Result<(u32, Rc<RefCell<Contract>>)> {
```

Comment:
That would be my preference as well but I would have to do some more research on the standard JSON (AST) output and whether it changed in any way meaningful way across solc versions to better know how to model it and if there are any moving parts that we would need to take into account.

I'm happy to do that in a follow-up, if that's okay?

And re regression, I didn't touch the (de)serialization logic here but rather modified how we thread state when building the model, e.g. we construct upfront the sources (for the ownership changes outlined in the OP) and return the parsed contract rather than pass the mutable ID -> Contract mapping to the function and implicitly append it here, so the expects, unwraps and results should be the same as in #545 (so no regression per se).

The `catch_unwind` is just a safety precaution as I mentioned in https://github.com/NomicFoundation/edr/pull/614#discussion_r1738258449.

Reply:
Opened https://github.com/NomicFoundation/edr/issues/636 to track the deserialization improvements